lib/listen/listener.rb

```diff
@@ -37,73 +31,69 @@ class Listener
-        _info "Listen version: #{Listen::VERSION}"
-      end
-
-      @silencer = Silencer.new
-      _reconfigure_silencer({})
-
-      @tcp_mode = nil
-      if [:recipient, :broadcaster].include? args[1]
-        target = args.shift
-        @tcp_mode = args.shift
-        _init_tcp_options(target)
-      end
-
-      @directories = args.flatten.map { |path| Pathname.new(path).realpath }
-      @queue = Queue.new
-      @block = block
-      @registry = Celluloid::Registry.new
+    def initialize(*dirs, &block)
+      options = dirs.last.is_a?(Hash) ? dirs.pop : {}
 
-      transition :stopped
+      @config = Config.new(options)
+
+      eq_config = Event::Queue::Config.new(@config.relative?)
+      queue = Event::Queue.new(eq_config) { @processor.wakeup_on_event }
+
+      silencer = Silencer.new
+      rules = @config.silencer_rules
+      @silencer_controller = Silencer::Controller.new(silencer, rules)
+
+      @backend = Backend.new(dirs, queue, silencer, @config)
+
+      optimizer_config = QueueOptimizer::Config.new(@backend, silencer)
+
+      pconfig = Event::Config.new(
+        self,
+        queue,
+        QueueOptimizer.new(optimizer_config),
+        @backend.min_delay_between_events,
+        &block)
+
+      @processor = Event::Loop.new(pconfig)
+
+      super() # FSM
     end
 
     default_state :initializing
 
-    state :initializing, to: :stopped
-    state :paused, to: [:processing, :stopped]
+    state :initializing, to: :backend_started
+
+    state :backend_started, to: [:frontend_ready] do
+      backend.start
+    end
 
-    state :stopped, to: [:processing] do
-      _stop_wait_thread
-      if @supervisor
-        @supervisor.terminate
-        @supervisor = nil
-      end
+    state :frontend_ready, to: [:processing_events] do
+      processor.setup
     end
 
-    state :processing, to: [:paused, :stopped] do
-      if wait_thread # means - was paused
-        _wakeup_wait_thread
-      else
-        @last_queue_event_time = nil
-        _start_wait_thread
-        _init_actors
+    state :processing_events, to: [:paused, :stopped] do
+      processor.resume
+    end
 
-        # Note: make sure building is finished before starting adapter (for
-        # consistent results both in specs and normal usage)
-        sync(:record).build
+    state :paused, to: [:processing_events, :stopped] do
+      processor.pause
+    end
 
-        _start_adapter
-      end
+    state :stopped, to: [:backend_started] do
+      backend.stop # should be before processor.teardown to halt events ASAP
+      processor.teardown
     end
 
     # Starts processing events and starts adapters
     # or resumes invoking callbacks if paused
     def start
-      transition :processing
+      transition :backend_started if state == :initializing
+      transition :frontend_ready if state == :backend_started
+      transition :processing_events if state == :frontend_ready
+      transition :processing_events if state == :paused
```

Comment:
Because those steps are in a sequence - e.g. line 91 can set ':frontend_ready` and line 92 tests for that. We can't "jump" from `:processing` to `:paused`, because we have to go through the states in between (which is what's happening here). Perhaps the last line could be `if [:paused, :processing_events].include?(state)`.